### PR TITLE
Enhancements error viz

### PR DESCRIPTION
**Pull request by eleniv3d**
Still work in progress. I open it for discussion and in order to keep track of what's missing

+ [x] get rid of o_success for cvs exporter #86
+ [x] add x axis of graph for error visualization + unit #85 #81
+ [x] legend scale show decimals #84
+ [x] missing o in output of legend #80
+ [x] Allow the Viz component to show either beams-joints-assembly value graphs #88
-> implemented as a per_item toggle
+ [x] per face metric when analyzing joints #89
-> implemented if pcd input is already segmented per face

### CODE REVIEW

**Comment by DamienGilliard** [comment]
@eleniv3d Thanks ! I took the liberty of linking the issues, so they are closed automatically when merging.

I was just looking at the per-face metric issue. Do you see a clear way to solve it ? I'm just diving in and it's not 100% clear to me how we can do that the most efficiently.

**Comment by eleniv3d** [comment]
> @eleniv3d Thanks ! I took the liberty of linking the issues, so they are closed automatically when merging.
> 
> I was just looking at the per-face metric issue. Do you see a clear way to solve it ? I'm just diving in and it's not 100% clear to me how we can do that the most efficiently.

@DamienGilliard
I guess one way could be to already segment the pcd per face with a similar mechanism as for the joints. Then the pcd2mesh distance would automatically know that it is calculating for every face and will pop the corresponding message as is now the case for joints vs beams

The other way would be to specify that we want the calculation to happen per face with another toggle or ideally if we manage to make issue #116 work. but then again the distance calculation would look in the whole pcd of the joint.

**Comment by 9and3** [comment]
> @DamienGilliard
> I guess one way could be to already segment the pcd per face with a similar mechanism as for the joints. Perhaps the joint segmentator could already output clusters for faces directly if it doesn't add to the calculation time. Then the pcd2mesh distance would automatically know that it is calculating for every face and will pop the corresponding message as is now the case for joints vs beams

can we not use the DFFace's brep face instead of the entire brep when doing the calculation of the error @eleniv3d ? And repeat it for each DFFace. I have some doubts about having an extra clustering component just for faces.

**Comment by eleniv3d** [comment]
> @eleniv3d we do agree that you want a list of lists of pointclouds (and None if no pointcloud was found)? : [[face_1_joint_1, face_2_joint_1, ...], [face_1_joint_2, face_2_joint_2, ...], ..., [face_1_joint_n, face_2_joint_n...]] (Just double checking, I am the kind of person who messes this kind of things easily)

Sounds good!

**Comment by eleniv3d** [comment]
Possibly okay to review now. I struggled a bit with getting the desired data from the segmentators and in the end hacked it a bit. Maybe we can see together in the next meeting.

Notes

1. The input of the source point cloud can now be joints, beams or faces of any data structure. We internally flatten it and always compare it with the length of the corresponding lists from the assembly
2. We handle the exception if a pcd is invalid and output the same invalid pcd or color its "twin" mesh white
3. I'll do a separate PR for the CVS exporter

(per face calculation)
face

(invalid items visualization on pcd)
pcd

(invalid items visualization on mesh)
mesh

File: src/gh/components/DF_visualization/code.py [hunk 1]
```diff
@@ -41,12 +41,28 @@ def RunScript(self,
...
+                                                        multiple_curves = multiple_curves)
 
         return o_colored_geo, o_legend, o_histogram
+
+# if __name__ == "__main__":
+#     com = DFVisualization()
+#     o_colored_geo, o_legend, o_histogram = com.RunScript(
+#         i_result,
+#         i_viz_settings
+#         )
```
Comment (9and3) [inline src/gh/components/DF_visualization/code.py]: to remove

File: src/gh/components/DF_visualization/code.py [hunk 2]
```diff
@@ -1,20 +1,20 @@
...
 from diffCheck.df_visualization import DFVizSettings
 from diffCheck.df_error_estimation import DFVizResults
 from diffCheck import diffcheck_bindings
 
+
 class DFVisualization(component):
     def RunScript(self,
                   i_result: DFVizResults,
                   i_viz_settings: DFVizSettings):
+
```
Comment (9and3) [inline src/gh/components/DF_visualization/code.py:17]: the same we could add:
```py
if i_result or i_viz_settings:
    return None, None
```

**Review by 9and3 — APPROVED** [review]
hello @eleniv3d ! Thanks for this PR, I just left some comments and did some tiny cleaning myself! Merging this..

**Pull request merged**